### ISSUE TITLE
Clean-up Pod-Html:  a battle plan

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#18894]
In this ticket I describe a plan to refactor code under `ext/Pod-Html` in the Perl 5 core distribution.

**Rationale**

When, up until October 2019, we used rt.perl.org as our bug tracker, we used to group tickets by creating "Meta" tickets.  Around 2012 Ricardo Signes created one such Meta ticket called [Pod::Html / pod2html is a mess](https://rt.perl.org/Ticket/Display.html?id=112014) [GH 12019](https://github.com/Perl/perl5/issues/12019).  It was then and still is.  The purpose of this refactoring is to significantly reduce the amount of that mess.

There are two major aspects to this messiness.

First, the code in `ext/Pod-Html/lib/Pod/Html.pm` is not fully encapsulated.  State is held in a total of 19 lexical variables declared outside the scope of the main function, `pod2html()`.

Second, the tests in `ext/Pod-Html/t/*.t` cannot be run in parallel.  That's because the test suite is currently set up to create temporary files underneath `ext/Pod-Html` rather than true tempdirs.  This leads to race conditions.  We currently have a workaround in `t/harness` to require the tests in the `ext/Pod-Html/t` to be run one at a time.

**Overall Game Plan**

The two aspects of Pod-Html's messiness should be tackled in reverse order.

First, revise the test suite so that the tests can be run in parallel.

Then encapsulate the code in the main module.  Instead of state being held in lexical variables, hold it in a Pod::Html object.  Read POD and user configuration choices into that object; parse it within that object; compose HTML within that object; write HTML files from within that object.

We won't attempt to fix every existing deficiency in Pod-Html in this project, but we'll be in a much better position to accept pull requests for those deficiencies once the code and tests are cleaner.

Furthermore, in this project we are not attempting to change what Pod-Html actually does.  More specifically, accept for correction of bugs the `pod2html` utility should work exactly the same after this cleanup as before.  But, once again, we'll be in a better position to change the way that utility works going forward.

**Recommendations**

When you are refactoring messy code you are likely to end up with an enormous diff even if, in the end, the functional changes you are making are few.  That will be the case here.  Hence I plan to present this project as a series of **five** pull requests corresponding to five nested branches in my own fork of Perl 5 on github.  The pull requests will implement the following "order of battle":

1. Restructure the test suite to have each test files run in its own tempdir, thereby avoiding race conditions and addressing [some anomalous results](https://github.com/Perl/perl5/issues/12271).

2. Partial refactoring of `lib/Pod/Html.pm`:  Most state moved to a hash (which will later become a Pod::Html object).  Begin isolating portions of `pod2html()` into individual subroutines for readability and refactorability.  Auxiliary subroutines moved to a separate package, then imported back into main module and tests.

This second p.r. will require a change in Pod::Simple, which is maintained upstream on CPAN.  That change has already been made in Pod-Simple's [repository](https://github.com/perl-pod/pod-simple/commit/16265391350539fc9b0e64cb224f1bb328ca8778) but has not yet been released to CPAN.

3. Subroutines in `lib/Pod/Html.pm` re-ordered in order in which they're called.

4. Cleanup of subroutine for processing command-line options passed to `pod2html` utility.

5. All state now moved to a Pod::Html object.  `pod2html()` function is now a wrapper around a constructor for that object and a series of method calls.

I will, at least initially, post these p.r.s one at a time.  But if you want to see what the five steps will ultimately add up to, go to https://github.com/jkeenan/perl5/pull/7.

**What These Proposed Changes Will *Not* Do**

As mentioned above, these changes will not meaningfully change what module Pod::Html and utility `pod2html` currently do.

These changes will not address the question of whether Pod-Html should be dual-lifed, whether blead-upstream or cpan-upstream.  If Pod-Html is ultimately to get a CPAN release it will need a maintainer, but that maintainer will not be me.

These changes will make no judgment as to whether the HTML generated by `pod2html` is "good" or "modern" HTML.  They should make Pod::Html subclassable, which may facilitate improvements in the generated HTML (see, *e.g.,* https://github.com/Perl/perl5/issues/17986 and https://github.com/Perl/perl5/issues/18487).

**How I Prefer These Changes Be Reviewed**

My impression is that many people perform code review in the GitHub GUI in a way that generates a flood of unhelpful email messages.  Upon learning of a new pull request, they go to the GUI, click next to individual lines of code upon which they wish to comment, and start a "conversation" for each line.  Each such conversation arrives in one's Inbox as a separate message, usually without any context -- meaning that you have to go from your email client back to your browser to understand the comment.  After more rounds back and forth like this, you (*i.e.,* the person originally submitting the pull request) has to "resolve" the conversation -- otherwise upon merge GH will complain that a merge is being attempt before all conversations have been resolved.

I find this style of code review annoying in the extreme.  Apart from the fact that it generates unnecessary email, it often leaves one with only a fuzzy impression of what the reviewer thought of the code *overall.*

Fortunately, there is an alternative.

@rjbs did a review of an earlier version of this project in which he got both comments on individual lines and an overall assessment in one fell swoop -- one email, that is.  I'll paraphrase how he told me he did it:

> I clicked on the "Files changed" tab of the pull request and begun to scroll through the whole PR.  When I saw a line on which I wanted to make a flippant remark, I clicked the blue ➕ to the left of the line number and the text area appeared.
> 
> After writing my witty line, I click "Start a review."  This begins buffering any subsequent remarks I make until, when done, I click "Submit review."  That button will replace the "Review changes" button at the top right when I've begun a review.
> 
> All the while, I can click "add single comment" if I see something I want to call out separately.
> 

So, please try to consolidate any comments you have into a single message per pull request (at least initially).

**Historical Note and Acknowledgements**

This work actually began at a hackathon sponsored by DuckDuckGo in Paoli, Pennsylvania in June 2012.  I was looking for a project and Rik Signes suggested I look into the possibility of making Pod-Html dual-life.  I [explored that possibility](https://github.com/jkeenan/Pod-Html) but the messiness defeated me.

Earlier versions of this project have benefited from feedback from Rik Signes, Nicholas Clark and Brett Estrade.  Craig Berry provided one patch for correct functioning on VMS.

**Comment by jkeenan** [comment]
> In this ticket I describe a plan to refactor code under `ext/Pod-Html` in the Perl 5 core distribution.
> [snip]

> 
> When you are refactoring messy code you are likely to end up with an enormous diff even if, in the end, the functional changes you are making are few. That will be the case here. Hence I plan to present this project as a series of **five** pull requests corresponding to five nested branches in my own fork of Perl 5 on github. The pull requests will implement the following "order of battle":
> 
>     1. Restructure the test suite to have each test files run in its own tempdir, thereby avoiding race conditions and addressing [some anomalous results](https://github.com/Perl/perl5/issues/12271).
> 

This code was merged to blead today in p.r. 18895.  Next up will be:

>     2. Partial refactoring of `lib/Pod/Html.pm`:  Most state moved to a hash (which will later become a Pod::Html object).  Begin isolating portions of `pod2html()` into individual subroutines for readability and refactorability.  Auxiliary subroutines moved to a separate package, then imported back into main module and tests.
> 
> 
> This second p.r. will require a change in Pod::Simple, which is maintained upstream on CPAN. That change has already been made in Pod-Simple's [repository](https://github.com/perl-pod/pod-simple/commit/16265391350539fc9b0e64cb224f1bb328ca8778) but has not yet been released to CPAN.
>

**Comment by jkeenan** [comment]
> > In this ticket I describe a plan to refactor code under `ext/Pod-Html` in the Perl 5 core distribution.
> > [snip]
> 
> > When you are refactoring messy code you are likely to end up with an enormous diff even if, in the end, the functional changes you are making are few. That will be the case here. Hence I plan to present this project as a series of **five** pull requests corresponding to five nested branches in my own fork of Perl 5 on github. The pull requests will implement the following "order of battle":
> > ```
> > 1. Restructure the test suite to have each test files run in its own tempdir, thereby avoiding race conditions and addressing [some anomalous results](https://github.com/Perl/perl5/issues/12271).
> > ```
> 
> This code was merged to blead today in p.r. 18895. Next up will be:
> 
> > ```
> > 2. Partial refactoring of `lib/Pod/Html.pm`:  Most state moved to a hash (which will later become a Pod::Html object).  Begin isolating portions of `pod2html()` into individual subroutines for readability and refactorability.  Auxiliary subroutines moved to a separate package, then imported back into main module and tests.
> > ```
> >   
> > This second p.r. will require a change in Pod::Simple, which is maintained upstream on CPAN. That change has already been made in Pod-Simple's [repository](https://github.com/perl-pod/pod-simple/commit/16265391350539fc9b0e64cb224f1bb328ca8778) but has not yet been released to CPAN.

@khwilliamson released Pod-Simple-3.43 to CPAN, clearing the way for the 2nd of 5 pull requests.  Following review by @rjbs, this p.r. was merged into blead this evening.

Next up:

3. Subroutines in lib/Pod/Html.pm re-ordered in order in which they're called.

This will be handled in pull request https://github.com/Perl/perl5/pull/18963.

**Comment by jkeenan** [comment]
The 5 pull requests that were anticipated in response to this issue have all been merged into blead.  This ticket is now closable.  Over the course of the coming week, I will (a) create new issues for all the suggestions made during review of these p.r.s which were deferred until after completion of the project; and (b) review the status of all existing issues and p.r.s with the 'ext-Pod-Html' label.

Thank you very much.
Jim Keenan
